src/core/opamStd.ml

```diff
@@ -920,50 +931,119 @@ module OpamSys = struct
     else
       fun x -> x
 
-  let guess_shell_compat () =
-    let parent_guess =
-      if Sys.win32 then
-        None
+  let windows_max_ancestor_depth = 5
+
+  (** [windows_ancestor_process_names] finds the names of the parent of the
+      current process and all of its ancestors up to [max_ancestor_depth]
+      in length.
+
+      The immediate parent of the current process will be first in the list.
+    *)
+  let windows_ancestor_process_names () =
+    let rec helper pid depth =
+      if depth > windows_max_ancestor_depth then []
       else
-        let ppid = Unix.getppid () in
-        let dir = Filename.concat "/proc" (string_of_int ppid) in
+      try
+        OpamStubs.(getProcessName pid ::
+                   helper
+                     (getParentProcessID pid)
+                     (depth + 1))
+      with Failure _ -> []
+    in
+    lazy (
+      try
+        let parent = OpamStubs.getCurrentProcessID () in
+        helper (OpamStubs.getParentProcessID parent) 0
+      with Failure _ -> []
+    )
+
+  type shell_choice = Reject | Accept of shell
+
+  let windows_get_shell =
+    let categorize_process = function
+      | "powershell.exe" | "powershell_ise.exe" -> Some (Accept SH_win_powershell)
+      | "pwsh.exe" -> Some (Accept SH_pwsh)
+      | "cmd.exe" -> Some (Accept SH_win_cmd)
+      | "env.exe" ->
+        (* If the nearest ancestor is env.exe it may be `env bash` or
+           even `env cmd.exe`. On Windows we can't see whether bash,
+           cmd or something else was chosen by `env ...`. So kick
+           out of categorization immediately. *)
```

Comment:
On my system, if I do `env ./opam.exe` (where opam.exe is _native_) then I do indeed see `env` as the ancestor with no further information. However if I do, say, `env bash -c './opam.exe'` then I get `bash.exe` as the immediate parent.

Regardless, if we see `env`, we're not being very Windows-y here - would it better to return `Some (Accept SH_sh)` here? I don't know how that interacts with Diskuv's wrappers, but with `Reject`, I get PowerShell output in Cmd if I run `C:\cygwin64\bin\env ./opam.exe env` or similarly in Cygwin's `bash` if I run `SHELL= env ./opam.exe env` but it seems more logical if this were to return `sh`-style environment output in both those contexts.

Reply:
After the original PR was written I got rid of using `env` to launch certain forms of Opam. The commentary and the Opam exceptions are at

https://github.com/diskuv/dkml-runtime-apps/blob/cac99123caa771770ba129cb80fbbbb0b986e9f1/src/with-dkml/cmdline.ml#L56-L91

Basically, `opam env` and `opam switch *` do not use `env.exe`; every other Opam invocation uses `env.exe` because it sets+configures the pager so `opam ... --help` works and probably other things I didn't capture.

So I'm fine with `sh`-style output for `env`. Honestly, the complexity has just been lifted from Opam to the wrapper though. If you are fine with that I'll change to `sh`-style for `env`

Reply:
That makes sense, thanks. Let's go with the `sh`-style for `env`, therefore, please.

Reply:
Done